chore(frontend): uplift Cypress to v13

hainenber opened the pull request:
<!---
Please write the PR title following the conventions at https://www.conventionalcommits.org/en/v1.0.0/
Example:
fix(dashboard): load charts correctly
-->
chore(frontend): uplift Cypress to v13


### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
Uplift Cypress to currency (v13). Acceptance threshold is all Cypress tests in CI pass with flying colors :D

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
<!--- Skip this if not applicable -->
N/A

### TESTING INSTRUCTIONS
<!--- Required! What steps can be taken to manually verify the changes? -->

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

hainenber committed:
feat(frontend): uplift Cypress to v11

Signed-off-by: hainenber <dotronghai96@gmail.com>

hainenber committed:
fix(ci): uplift Cypress to v12 to fix failed post-v11 E2E failure

Signed-off-by: hainenber <dotronghai96@gmail.com>

hainenber committed:
fix(ci): remove obsolete workaround for Cypress <12.15.0

Signed-off-by: hainenber <dotronghai96@gmail.com>

hainenber committed:
chore(build): disable interruptive Cypress 12's ESLint rule

Signed-off-by: hainenber <dotronghai96@gmail.com>

hainenber committed:
feat(build): uplift Cypress to v13 to fix incompatibility with Webpack 5

Signed-off-by: hainenber <dotronghai96@gmail.com>

hainenber committed:
fix(ci): preserve cookie across test for Cypress v12

Signed-off-by: hainenber <dotronghai96@gmail.com>

hainenber committed:
chore: fix ESlint issue

Signed-off-by: hainenber <dotronghai96@gmail.com>

hainenber committed:
fix(e2e): fix failed E2E test in Cypress 13 by disabling test isolation

Signed-off-by: hainenber <dotronghai96@gmail.com>

hainenber committed:
chore(e2e): remove deprecated Cypress config

Signed-off-by: hainenber <dotronghai96@gmail.com>

hainenber committed:
fix(e2e): fix leftover failed E2E tests

Signed-off-by: hainenber <dotronghai96@gmail.com>

rusackas commented:
Hey, thanks for getting into this! This'll enable some cool stuff, including test playback on the Cypress dashboard, which is pretty nifty!

hainenber committed:
fix(e2e): dedicated test setup for post-Cypress13's Bar Chart v2

The coordinates to drill the BarChartV2 are much harder due to slender visualization so we have to increase the `y` coordinate to get more chance clicking into correct areas.

This in turns forces the test assertion to be more laxed to handle many states the drill-by can have.

Signed-off-by: hainenber <dotronghai96@gmail.com>

hainenber committed:
fix(e2e): resolve failed `GaugeChart` test after upgrading Cypress to v13

Signed-off-by: hainenber <dotronghai96@gmail.com>

hainenber committed:
fix(e2e): resolve failed `nativeFilters` test after upgrading Cypress to v13

Signed-off-by: hainenber <dotronghai96@gmail.com>

hainenber commented on superset-frontend/cypress-base/cypress/e2e/dashboard/drillby.test.ts:
```diff
@@ -524,10 +524,141 @@ describe('Drill by modal', () => {
         [70, 94],
-        [362, 68],
-      ]);
+        [362, 474],
```
I'm not sure if it's due to Cypress upgrade or else but the previous coordinates were flaky. I have to manually tinker around to pick out the least "flaky" coordinates to pass local CI

sadpandajoe replied:
Yeah coordinates are always flakey in cypress as things can always add or subtract to it. You could try writing a function where you can pass in what you are looking for and then it'll generate an approximate window coordinate for you. I have that in our repo and it's still flakey but i'm not adding prs to make these updates as much

hainenber commented on superset-frontend/cypress-base/cypress/e2e/dashboard/drillby.test.ts:
```diff
@@ -588,7 +719,7 @@ describe('Drill by modal', () => {
       testEchart('gauge_chart', 'Gauge Chart', [
         [151, 95],
-        [300, 143],
+        [300, NaN], // this fixes the test after upgrading to Cypress 13 
```
This is completely voodoo magic to me. Previous coordinates didn't work and frustratingly I entered `NaN` and it worked

hainenber commented on superset-frontend/cypress-base/cypress/e2e/dashboard/nativeFilters.test.ts:
```diff
@@ -428,14 +428,13 @@ describe('Native filters', () => {
       cy.reload();
       cy.get(dataTestChartName(testItems.topTenChart.name)).within(() => {
         cy.contains(testItems.filterDefaultValue).should('be.visible');
-        cy.contains(testItems.filterOtherCountry).should('not.exist');
```
`cy.reload()` in Cypress 13 seems to completely remove local cache for stored filter inputs. Lmk if my change here is a fix or a band-aid :D

hainenber committed:
chore: fix ESLint error

Signed-off-by: hainenber <dotronghai96@gmail.com>

hainenber committed:
fix(e2e): prevent renderer crash due to memory overconsumption

Signed-off-by: hainenber <dotronghai96@gmail.com>

hainenber committed:
Revert "fix(e2e): prevent renderer crash due to memory overconsumption"

This reverts commit db4985ab5c805262e2e7547705daa887cf84833e.

hainenber committed:
fix(e2e): increase retries for flaky test + skip a weirdly running `nativeFilters` test

Signed-off-by: hainenber <dotronghai96@gmail.com>

hainenber committed:
chore: execute `npm run lint-fix`

Signed-off-by: hainenber <dotronghai96@gmail.com>

hainenber committed:
fix(e2e): prevent renderer crash due to memory overconsumption

Signed-off-by: hainenber <dotronghai96@gmail.com>

sadpandajoe commented on superset-frontend/cypress-base/cypress.config.ts:
```diff
@@ -37,30 +36,10 @@ export default eyesPlugin(
       openMode: 0,
     },
     e2e: {
+      experimentalMemoryManagement: true,
```
I see we are doing `{ testIsolation: false }` in the describe blocks, is there a reason to not just add that to the config instead?

hainenber replied:
It's due to [default test isolation](https://docs.cypress.io/guides/references/migration-guide#Test-Isolation) in Cypress 12. While there are many tests sharing context, as evidenced by the amount of `{ testIsolation: false }`, the remaining are isolated and should be encouraged instead.

We'll probably need to refactor these context-sharing tests later to be isolated in future PRs.

hainenber committed:
Merge branch 'master' into uplift-cypress-to-v13

Signed-off-by: hainenber <dotronghai96@gmail.com>

mistercrunch commented:
I'm generally bullish on landing this, despite the subtle flavoring of the voodoo before/after. Can't comment on specific issues here.

hainenber commented:
Closed due to inactivity. A Playwright migration PR could be better in terms of p/p
